## Part 1: Preprocessing

In [13]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df

Age Attrition     BusinessTravel              Department  \
0      41       Yes      Travel_Rarely                   Sales   
1      49        No  Travel_Frequently  Research & Development   
2      37       Yes      Travel_Rarely  Research & Development   
3      33        No  Travel_Frequently  Research & Development   
4      27        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
1465   36        No  Travel_Frequently  Research & Development   
1466   39        No      Travel_Rarely  Research & Development   
1467   27        No      Travel_Rarely  Research & Development   
1468   49        No  Travel_Frequently                   Sales   
1469   34        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                    1          2  Life Sciences                        2   
1                    8          1  Life Sciences                        3   
2                    2          2          Other                        4   
3                    3          4  Life Sciences                        4   
4                    2          1        Medical                        1   
...                ...        ...            ...                      ...   
1465                23          2        Medical                        3   
1466                 6          1        Medical                        4   
1467                 4          3  Life Sciences                        2   
1468                 2          3        Medical                        4   
1469                 8          3        Medical                        2   

      HourlyRate  JobInvolvement  ...  PerformanceRating  \
0             94               3  ...                  3   
1             61               2  ...                  4   
2             92               2  ...                  3   
3             56               3  ...                  3   
4             40               3  ...                  3   
...          ...             ...  ...                ...   
1465          41               4  ...                  3   
1466          42               2  ...                  3   
1467          87               4  ...                  4   
1468          63               2  ...                  3   
1469          82               4  ...                  3   

     RelationshipSatisfaction  StockOptionLevel TotalWorkingYears  \
0                           1                 0                 8   
1                           4                 1                10   
2                           2                 0                 7   
3                           3                 0                 8   
4                           4                 1                 6   
...                       ...               ...               ...   
1465                        3                 1                17   
1466                        1                 1                 9   
1467                        2                 1                 6   
1468                        4                 0                17   
1469                        1                 0                 6   

      TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                         0               1               6   
1                         3               3              10   
2                         3               3               0   
3                         3               3               8   
4                         3               3               2   
...                     ...             ...             ...   
1465                      3               3               5   
1466                      5               3               7   
1467                      0               3               6   
1468                      3               2               9   
1469                      3               4  

In [2]:
# Determine the number of unique values in each column
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [4]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]


In [5]:
# Create a list of at least 10 column names to use as X data
X_columns = ['Age', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'NumCompaniesWorked', 'TotalWorkingYears']


# Create X_df using your selected columns
x_df = attrition_df[X_columns]


# Show the data types for X_df
x_df.dtypes


Age                        int64
DistanceFromHome           int64
Education                  int64
EnvironmentSatisfaction    int64
HourlyRate                 int64
JobInvolvement             int64
JobLevel                   int64
JobSatisfaction            int64
NumCompaniesWorked         int64
TotalWorkingYears          int64
dtype: object

In [6]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split


In [9]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
x_df = x_df.apply(pd.to_numeric, errors='coerce')



In [10]:
# Create a StandardScaler
scaler = StandardScaler()


# Fit the StandardScaler to the training data
scaler.fit(x_df)


# Scale the training and testing data
x_df = scaler.transform(x_df)


In [12]:
from sklearn.preprocessing import OneHotEncoder

# Create a OneHotEncoder for the Department column
encoder = OneHotEncoder(sparse_output=False, drop='first')


# Fit the encoder to the training data
encoder.fit(y_df[['Department']])


# Create two new variables by applying the encoder
# to the training and testing data
y_df_encoded = encoder.transform(y_df[['Department']])
y_df_encoded = pd.DataFrame(y_df_encoded, columns=encoder.get_feature_names_out(['Department']))


In [15]:
# Create a OneHotEncoder for the Attrition column
encoder = OneHotEncoder(sparse_output=False, drop='first')


# Fit the encoder to the training data
encoder.fit(y_df[['Attrition']])


# Create two new variables by applying the encoder
# to the training and testing data
y_df_encoded2 = encoder.transform(y_df[['Attrition']])
y_df_encoded2 = pd.DataFrame(y_df_encoded2, columns=encoder.get_feature_names_out(['Attrition']))
# show array
y_df_encoded2.head()


Attrition_Yes
0            1.0
1            0.0
2            1.0
3            0.0
4            0.0

## Part 2: Create, Compile, and Train the Model

In [16]:
# Find the number of columns in the X training data.
num_cols = x_df.shape[1]


# Create the input layer
input_layer = layers.Input(shape=(num_cols,))


# Create at least two shared layers
layer1 = layers.Dense(64, activation='relu')(input_layer)
layer2 = layers.Dense(32, activation='relu')(layer1)


In [17]:
# Create a branch for Department
# with a hidden layer and an output layer
branch1 = layers.Dense(16, activation='relu')(layer2)

# Create the hidden layer
branch1 = layers.Dense(8, activation='relu')(branch1)


# Create the output layer
branch1 = layers.Dense(y_df_encoded.shape[1], activation='softmax', name='Department')(branch1)



In [18]:
# Create a branch for Attrition
# with a hidden layer and an output layer
branch2 = layers.Dense(16, activation='relu')(layer2)

# Create the hidden layer
branch2 = layers.Dense(8, activation='relu')(branch2)


# Create the output layer
branch2 = layers.Dense(y_df_encoded2.shape[1], activation='softmax', name='Attrition')(branch2)


In [19]:
# Create the model
model = Model(inputs=input_layer, outputs=[branch1, branch2])


# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


# Summarize the model
model.summary()


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 10)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 64)        │        704 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 32)        │      2,080 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 16)        │        528 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_4 (Dense)     │ (None, 16)        │        528 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_3 (Dense)     │ (None, 8)         │        136 │ dense_2[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_5 (Dense)     │ (None, 8)         │        136 │ dense_4[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department (Dense)  │ (None, 2)         │         18 │ dense_3[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition (Dense)   │ (None, 1)         │          9 │ dense_5[0][0]     │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 4,139 (16.17 KB)

 Trainable params: 4,139 (16.17 KB)

 Non-trainable params: 0 (0.00 B)

In [21]:
# Compile the model with metrics for each output
model.compile(
    optimizer='adam',
    loss=['categorical_crossentropy', 'categorical_crossentropy'],  # Loss for each output
    metrics=[['accuracy'], ['accuracy']]  # Metrics for each output
)

# Train the model
history = model.fit(
    x_df, 
    [y_df_encoded, y_df_encoded2], 
    epochs=50, 
    batch_size=32, 
    validation_split=0.2
)

Epoch 1/50


C:\Users\mrell\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\ops\nn.py:908: UserWarning: You are using a softmax over axis -1 of a tensor of shape (None, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(
C:\Users\mrell\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\losses\losses.py:33: SyntaxWarning: In loss categorical_crossentropy, expected y_pred.shape to be (batch_size, num_classes) with num_classes > 1. Received: y_pred.shape=(None, 1). Consider using 'binary_crossentropy' if you only have 2 classes.
  return self.fn(y_true, y_pred, **self._fn_kwargs)


17/37 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - Attrition_accuracy: 0.1756 - Attrition_loss: 0.0000e+00 - Department_accuracy: 0.3213 - Department_loss: 0.7009 - loss: 0.7009 

C:\Users\mrell\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\ops\nn.py:908: UserWarning: You are using a softmax over axis -1 of a tensor of shape (None, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(
C:\Users\mrell\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\losses\losses.py:33: SyntaxWarning: In loss categorical_crossentropy, expected y_pred.shape to be (batch_size, num_classes) with num_classes > 1. Received: y_pred.shape=(None, 1). Consider using 'binary_crossentropy' if you only have 2 classes.
  return self.fn(y_true, y_pred, **self._fn_kwargs)


37/37 ━━━━━━━━━━━━━━━━━━━━ 3s 12ms/step - Attrition_accuracy: 0.1720 - Attrition_loss: 0.0000e+00 - Department_accuracy: 0.4015 - Department_loss: 0.6881 - loss: 0.6881 - val_Attrition_accuracy: 0.1463 - val_Attrition_loss: 0.0000e+00 - val_Department_accuracy: 0.6565 - val_Department_loss: 0.6384 - val_loss: 0.6326
Epoch 2/50
37/37 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - Attrition_accuracy: 0.1464 - Attrition_loss: 0.0000e+00 - Department_accuracy: 0.6806 - Department_loss: 0.6347 - loss: 0.6347 - val_Attrition_accuracy: 0.1463 - val_Attrition_loss: 0.0000e+00 - val_Department_accuracy: 0.6701 - val_Department_loss: 0.6194 - val_loss: 0.6184
Epoch 3/50
37/37 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - Attrition_accuracy: 0.1841 - Attrition_loss: 0.0000e+00 - Department_accuracy: 0.7100 - Department_loss: 0.5990 - loss: 0.5990 - val_Attrition_accuracy: 0.1463 - val_Attrition_loss: 0.0000e+00 - val_Department_accuracy: 0.6701 - val_Department_loss: 0.6055 - val_loss: 0.6032
Epoch 4/50
37/37 ━━━━━━━━━

In [23]:
# Evaluate the model with the testing data
eval_results = model.evaluate(x_df, [y_df_encoded, y_df_encoded2], verbose=0)
# Show the evaluation results
print(f"Evaluation results: {eval_results}")
# Create a DataFrame with the training history
history_df = pd.DataFrame(history.history)


Evaluation results: [176.3953399658203, 176.34034729003906, 0.0, 0.16122448444366455, 0.3034013509750366]


In [24]:
# Print the accuracy for both department and attrition
print(f"Training accuracy for Department: {history_df['Department_accuracy'].iloc[-1]}")
print(f"Training accuracy for Attrition: {history_df['Attrition_accuracy'].iloc[-1]}")


Training accuracy for Department: 0.5688775777816772
Training accuracy for Attrition: 0.16496598720550537


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy ain't great here. Attrition’s stuck at ~0.16, probably 'cause the data’s imbalanced. Stuff like F1-score or AUC-ROC would tell us more about nailing the minority class.


2. I used softmax for both, but that’s messing up Attrition—it’s binary, so sigmoid’s the way to go. Department’s multi-class, so softmax is cool there.


3. Fix Attrition to sigmoid with binary_crossentropy, handle imbalanced data with class weights or SMOTE, impute NaNs in x_df, scale features, tweak layers or add dropout, and use early stopping to avoid overfitting.

